### PR TITLE
Highlight insecure element and scroll to it

### LINKED ISSUE #9: Visually highlight insecure assets in the editor
It'd be a nice UX touch to visually highlight to the assets that are flagged as insecure. Ideally, this would add a border to assets in the visual editor, and highlight the asset text in the text editor.

https://cl.ly/0Y1x1C3i271L

### PR DESCRIPTION
**Pull request by cadic**
### Description of the Change

Adds "View element" button, which scrolls editor to the element with insecure content. The element is highlighted with the red outline.

Works both in Classic and Block editor.

The outline is cleared before every new content check, post save or preview.

Closes #9

### Verification Process

1. Create new post
2. Add some content and put insecure element/block at the end of the content (outside of the first screen)
3. Try to save the post, the insecure warning should appear
4. Click on "View element" link/button (depends on the Block or Classic editor you use)
5. The page expected to scroll to the element/block and add a red outline to it

https://user-images.githubusercontent.com/288381/170232262-0026ef52-7567-45cd-aae0-494d78b81250.mp4

### Possible Drawbacks

The original tracking design doesn't allow to identify exact element/block which refers to insecure content. We only have list of URLs, which is used in this PR to find elements in the editor.

If the content have multiple elements with the same URL, multiple items with this URL will be added to the warning block. But any "View element" link from the identical URLs will only scroll/highlight the first element.

### Checklist:

- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests passed.

### Changelog Entry

> Added - "View element" link to highlight and auto-scroll to the insecure element
> Fixed - Dependencies of JavaScript assets

### Credits

Props @cadic, @psorensen, @adamsilverstein, @dkotter

### CODE REVIEW

File: src/js/gutenberg.js
Comment (peterwilsoncc) [inline src/js/gutenberg.js]: There are a few items in use in this file that are not listed in the dependencies (jQuery, apiRequest, domReady). Some of these are legacy issues but it would be good to add them to ensure this update doesn't end up changing the order in which files are included.

https://github.com/10up/insecure-content-warning/blob/42a8fbf51fd9ef62afb28c3070506143cd1b0a5b/includes/assets.php#L34-L40
:1234: To avoid a duplicate comment, are you able to review both JS files dependencies?

https://github.com/10up/insecure-content-warning/blob/42a8fbf51fd9ef62afb28c3070506143cd1b0a5b/includes/assets.php#L60-L66

File: includes/assets.php
Comment (peterwilsoncc) [inline includes/assets.php:125]:
```suggestion
	if ( empty( $mce_css ) ) {
		return $url;
	}
	$mce_css = $url . ',' . $mce_css;
```

**Review by peterwilsoncc — COMMENTED** [review]
While testing, I notice the insecure element only has the border added if the "scroll to element" button is clicked. Is this intentional?

red-borders

(Gif only shows block editor but the same happens in the classic editor)


🔢 Is used to indicate the comment applies elsewhere in the PR.

**Comment by cadic** [comment]
> While testing, I notice the insecure element only has the border added if the "scroll to element" button is clicked. Is this intentional?

From the original issue report and discussion, the highlight expected to happen on an explicit action. Thanks to your demo, it looks like we should avoid highlighting multiple items at a time, I will add "blur all" action before highlighting new element.

**Comment by cadic** [comment]
Added a note about multiple elements with the same URL to PR description

**Comment by peterwilsoncc** [comment]
The changes are working as expected, only one image is highlighted at a time.

I did notice in the block editor that the code will no longer scroll to an element if the post is unsaved. Again a gif to demonstrate:

The steps I am taking are:

1. Create a post
2. Add some content, insecure images
3. Click publish, warning show
4. Click view element -- scrolling fails
5. Click save draft
6. Click view element -- scrolling fails
7. Go to post list view
8. Reopen post
9. Click publish
10. Click view element -- scrolling succeeds 

unsaved-post

I am on leave next week so you'll need to reassign this issue for review.

**Comment by jeffpaul** [comment]
@cadic any thoughts on updating based on @peterwilsoncc's latest findings above?

**Comment by cadic** [comment]
@peterwilsoncc the last issue is fixed: now highlight works for all blocks, including just added.

Tested with:
- Image
- Cover
- Media and Text
- Audio
- Video

**Review by peterwilsoncc — COMMENTED** [review]
This is now working as expected, are you able to take a look at this earlier comment about dependencies to ensure all required files are listed for the JS files?

https://github.com/10up/insecure-content-warning/pull/73/files#r882190332

**Comment by cadic** [comment]
@peterwilsoncc I've added [Webpack Dependency Extractor Plugin](https://www.npmjs.com/package/@wordpress/dependency-extraction-webpack-plugin) to handle with dependencies

File: config/webpack.config.common.js
Comment (peterwilsoncc) [inline config/webpack.config.common.js]:
```suggestion
		new DependencyExtractorWebpackPlugin( {
			injectPolyfill: false,
			combineAssets: false,
			requestToExternal( request ) {
				if ( request === 'underscore' ) {
					return '_';
				}
			},
		} ),
```

I'm not sure why but the plugin doesn't include underscore so it's missing from the classic editor. This should fix it.

In Distributor I needed to set `injectPolyfill` to true but that might not be needed here.

**Review by peterwilsoncc — CHANGES_REQUESTED** [review]
Sorry Max, one final thing due to the dependency extractor not including underscore.

**Review by peterwilsoncc — DISMISSED** [review]
Thanks Max, I checked the built assets and they now include underscore.

The eslint and dependency failures look to be coming from elsewhere so I don't think they need to block merge.

**Comment by jeffpaul** [comment]
@dkotter will leave it up to you if you want to merge this in ahead of the 1.0.3 release or not

**Review by peterwilsoncc — APPROVED** [review]
Still happy with this.

For some reason the eslint and cypress jobs have started failing since I last looked at this repo. I did a quick test in another branch and it seems unrelated to this PR.

**Pull request merged**
